R/rectangle.R

```diff
@@ -167,104 +222,461 @@ hoist <- function(.data, .col, ..., .remove = TRUE, .simplify = TRUE, .ptype = l
+    abort("`indices_include` must be `NULL` or a single `TRUE` or `FALSE`.")
+  }
 
-  values_to <- values_to %||% col
-  if (!is.null(indices_to)) {
-    indices_include <- indices_include %||% TRUE
+  if (is.null(values_to)) {
+    values_to <- "{col}"
+  }
+  if (!is_string(values_to)) {
+    abort("`values_to` must be a single string or `NULL`.")
+  }
+
+  if (is.null(indices_to)) {
+    indices_to <- paste0(values_to, "_id")
   } else {
-    indices_to <- paste0(col, "_id")
+    if (is_false(indices_include)) {
+      abort("Can't set `indices_include` to `FALSE` when `indices_to` is supplied.")
+    }
+    indices_include <- TRUE
+  }
+  if (!is_string(indices_to)) {
+    abort("`indices_to` must be a single string or `NULL`.")
+  }
+
+  values_to <- glue_col_names(values_to, col_names)
+  values_to <- vec_recycle(values_to, size = n_col_names)
+
+  indices_to <- glue_col_names(indices_to, col_names)
+  indices_to <- vec_recycle(indices_to, size = n_col_names)
+
+  for (i in seq_along(cols)) {
+    col <- cols[[i]]
+    col_name <- col_names[[i]]
+    col_values_to <- values_to[[i]]
+    col_indices_to <- indices_to[[i]]
+
+    data[[col]] <- col_to_long(
+      col = data[[col]],
+      name = col_name,
+      values_to = col_values_to,
+      indices_to = col_indices_to,
+      indices_include = indices_include
+    )
+  }
+
+  data <- unchop(data, all_of(cols))
+
+  for (i in seq_along(cols)) {
+    col <- cols[[i]]
+
+    data[[col]] <- df_simplify(
+      x = data[[col]],
+      ptype = ptype,
+      transform = transform,
+      simplify = simplify
+    )
+  }
+
+  unpack(data, all_of(cols), names_repair = names_repair)
+}
+
+# Converts a column of any type to a `list_of<tbl>`
+col_to_long <- function(col,
+                        name,
+                        values_to,
+                        indices_to,
+                        indices_include) {
+  if (is.null(col)) {
+    abort(glue("Invalid `NULL` column detected for column `{name}`."))
   }
 
-  data[[col]] <- map(
-    data[[col]], vec_to_long,
-    col = col,
+  if (!vec_is_list(col)) {
+    ptype <- vec_ptype(col)
+    col <- tidyr_chop(col)
+    col <- new_list_of(col, ptype = ptype)
+  }
+
+  info <- collect_indices_info(col, indices_include)
+  indices_include <- info$indices_include
+  indices <- info$indices
+  index_ptype <- info$index_ptype
+
+  # If we don't have a list_of, then a `NULL` `col_ptype` will get converted to
+  # a 1 row, 1 col tibble for `elt_ptype`
+  col_ptype <- list_of_ptype(col)
+  elt_ptype <- elt_to_long(
+    x = col_ptype,
+    index = index_ptype,
+    name = name,
     values_to = values_to,
     indices_to = indices_to,
     indices_include = indices_include
   )
+  elt_ptype <- vec_ptype(elt_ptype)
+
+  # Avoid expensive dispatch from `[[.list_of`, and allow for `[[<-`
+  col <- tidyr_new_list(col)
+
+  for (i in seq_along(col)) {
+    col[[i]] <- elt_to_long(
+      x = col[[i]],
+      index = indices[[i]],
```

Comment:
Can't use `map2()` here because `indices` might be `NULL`, but `NULL[[i]]` is valid and just returns `NULL` so the for loop form worked well